high_performance_computing/hpc_openmp/04_synchronisation.md

```diff
@@ -0,0 +1,529 @@
+conflicts caused by multiple threads accessing the same piece of shared data. It's important to synchronise threads
+properly so they are able to work together, and not interfere with data another thread is accessing or modifying.
+Synchronisation is also important for data dependency, to make sure that a thread has access to the data it requires.
+This is particularly important for algorithms which are iterative.
+
+The synchronisation mechanisms in OpenMP are incredibly important tools, as they are used to avoid race conditions. Race
+conditions are have to be avoided, otherwise they *can* result in data inconsistencies if we have any in our program. A
+race  condition happens when two, or more, threads access and modify the same piece of data at the same time. To
+illustrate this, consider the diagram below:
+
+![Race conditions](fig/race_condition.png)
+
+Two threads access the same shared variable (with the value 0) and increment it by 1. Intuitively, we might except the
+final value of the shared variable to be 2. However, due to the potential for concurrent access and modification, we
+can't actually guarantee what it will be. If both threads access and modify the variable concurrently, then the
+final value will be 1. That's because both variables read the initial value of 0, increment it by 1, and write to the
+shared variable.
+
+In this case, it doesn't matter if the variable update does or doesn't happen concurrently. The inconsistency stems from
+the value initially read by each thread. If, on the other hand, one thread manages to access and modify the variable
+before the other thread can read its value, then we'll get the value we expect (2). For example, if thead 0 increments
+the variable before thread 1 reads it, then thread 1 will read a value of 1 and increment that by 1 givusing us the
+correct value of 2. This illustrates why it's called a race condition, because threads race each other to access and
+modify variables before another thread can!
+
+:::callout
+## Analogy: Editing a document
+
+Imagine two people trying to update the same document at the same time. If they don't communicate what they're doing,
+they might edit the same part of the document and overwrite each others changes, ending up with a messy and
+inconsistent document (e.g. when they come to merge changes later). This is just like what happens with a race
+condition in OpenMP. Different threads accessing and modifying the same part of memory, which results in messy and
+inconsistent memory operations and probably an incorrect result.
+:::
+
+::::challenge{title="Identifying race conditions"}
+Take a look at the following code example. What's the output when you compile
+and run this program? Where do you think the race condition is?
+
+```c
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#define NUM_TIMES 10000
+
+int main(void) {
+    int value = 0;
+
+#pragma omp parallel for
```

Comment:
should this pragma be lined up with the for loop?

Reply:
Fixed throughout - including those below!